File: templates/typescript-express/app/apis/bar.ts
/**
 * Example BYOF (Bring Your Own Framework) Express app
 *
 * This file demonstrates how to use Express with MooseStack for consumption
 * APIs using the ByofApi class.
 */

import express, { Request, Response, NextFunction } from "express";
import {
  sql,
  getMooseClients,
  ByofApi,
  type MooseClient,
} from "@514labs/moose-lib";
import { BarPipeline } from "../ingest/models";
import { Api, MooseCache } from "@514labs/moose-lib";
import { BarAggregatedMV } from "../views/barAggregated";
import { tags } from "typia";

interface QueryParams {
  maxResults: number;
}

// Extend Express Request to include our client
interface RequestWithClient extends Request<{}, {}, {}, QueryParams> {
  client?: MooseClient;
}

interface ResultItem {
  primaryKey: string;
  textLength: number;
}

/**
 * Creates and configures the Express app.
 * This function is called by the ByofApi class when the app is initialized.
 */
async function createExpressApp() {
  const app = express();
  const { client } = await getMooseClients();

  // Middleware to attach Moose client to requests
  app.use((req: RequestWithClient, _res: Response, next: NextFunction) => {
    req.client = client;
    next();
  });

  // Example: Custom endpoint that queries ClickHouse
  app.get("/moose", async (req: RequestWithClient, res: Response) => {
    const { client } = req;
    const SourceTable = BarPipeline.table!;
    const cols = SourceTable.columns;

    const query = sql` SELECT ${cols.primaryKey}, ${cols.textLength} FROM ${SourceTable} LIMIT ${req.query.maxResults || 10}`;

    // Set the result type to the type of the each row in the result set
    const resultSet = await client?.query.execute<ResultItem>(query);

    // Return the result set as an array of the result item type
    const data = await resultSet?.json();
    res.send(data);
  });

  return app;
}

/**
 * Register the Express app with Moose using the ByofApi class.
 * This follows the same pattern as Api and IngestApi classes.
 */
export const barByofApi = new ByofApi("bar-express-api", createExpressApp, {
  version: "1.0",
  metadata: { description: "Example Express app integrated with Moose" },
});

interface ApiQueryParams {
  orderBy?: "totalRows" | "rowsWithText" | "maxTextLength" | "totalTextLength";
  limit?: number;
  startDay?: number & tags.Type<"int32">;
  endDay?: number & tags.Type<"int32">;
}

interface ResponseData {
  dayOfMonth: number;
  totalRows?: number;
  rowsWithText?: number;
  maxTextLength?: number;
  totalTextLength?: number;
}

export const BarApi = new Api<ApiQueryParams, ResponseData[]>(
  "bar",
  async (
    { orderBy = "totalRows", limit = 5, startDay = 1, endDay = 31 },
    { client, sql },
  ) => {
    const cache = await MooseCache.get();
    const cacheKey = `bar:${orderBy}:${limit}:${startDay}:${endDay}`;

    // Try to get from cache first
    const cachedData = await cache.get<ResponseData[]>(cacheKey);
    if (cachedData && Array.isArray(cachedData) && cachedData.length > 0) {
      return cachedData;
    }

    const query = sql`
        SELECT 
          ${BarAggregatedMV.targetTable.columns.dayOfMonth},
          ${BarAggregatedMV.targetTable.columns[orderBy]}
        FROM ${BarAggregatedMV.targetTable}
        WHERE 
          dayOfMonth >= ${startDay} 
          AND dayOfMonth <= ${endDay}
        ORDER BY ${BarAggregatedMV.targetTable.columns[orderBy]} DESC
        LIMIT ${limit}
      `;

    const data = await client.query.execute<ResponseData>(query);
    const result: ResponseData[] = await data.json();

    await cache.set(cacheKey, result, 3600); // Cache for 1 hour

    return result;
  },
);

File: templates/python-webapp/app/apis/bar.py
"""
Example BYOF (Bring Your Own Framework) FastAPI app

This file demonstrates how to use FastAPI with MooseStack for consumption
APIs using the WebApp class.
"""
from fastapi import FastAPI, Request, HTTPException
from fastapi.responses import JSONResponse
from moose_lib.dmv2 import WebApp, WebAppConfig, WebAppMetadata
from moose_lib.dmv2.web_app_helpers import get_moose_utils
from app.db.views import BarAggregatedMV
from pydantic import BaseModel
from typing import Optional, List

app = FastAPI()


# Middleware to log requests
@app.middleware("http")
async def log_requests(request: Request, call_next):
    print(f"[bar.py] {request.method} {request.url.path}")
    response = await call_next(request)
    return response


# Health check endpoint
@app.get("/health")
async def health():
    """Health check endpoint"""
    return {
        "status": "ok",
        "timestamp": str(__import__("datetime").datetime.now()),
        "service": "bar-fastapi-api",
    }


# Query endpoint
@app.get("/query")
async def query(request: Request, limit: int = 10):
    """
    Query aggregated bar data.

    This endpoint demonstrates:
    - Accessing MooseStack utilities via get_moose_utils
    - Using the QueryClient to execute queries
    - Using the sql function for safe query building
    """
    moose = get_moose_utils(request)
    if not moose:
        raise HTTPException(
            status_code=500,
            detail="MooseStack utilities not available"
        )

    try:
        # Build the query
        query = f"""
            SELECT
                day_of_month,
                total_rows
            FROM BarAggregated
            ORDER BY total_rows DESC
            LIMIT {limit}
        """

        result = moose.client.query.execute(query)

        return {
            "success": True,
            "count": len(result),
            "data": result,
        }
    except Exception as error:
        print(f"Query error: {error}")
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


# POST endpoint for filtered data
class DataRequest(BaseModel):
    order_by: str = "total_rows"
    limit: int = 5
    start_day: int = 1
    end_day: int = 31


@app.post("/data")
async def data(request: Request, body: DataRequest):
    """
    Query aggregated bar data with filters.

    This endpoint demonstrates:
    - POST request handling
    - Request body validation with Pydantic
    - Dynamic query building based on request parameters
    """
    moose = get_moose_utils(request)
    if not moose:
        raise HTTPException(
            status_code=500,
            detail="MooseStack utilities not available"
        )

    try:
        # Validate order_by field
        valid_fields = ["total_rows", "rows_with_text", "max_text_length", "total_text_length"]
        if body.order_by not in valid_fields:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid order_by field: {body.order_by}"
            )

        query = f"""
            SELECT
                day_of_month,
                {body.order_by}
            FROM BarAggregated
            WHERE
                day_of_month >= {body.start_day}
                AND day_of_month <= {body.end_day}
            ORDER BY {body.order_by} DESC
            LIMIT {body.limit}
        """

        result = moose.client.query.execute(query)

        return {
            "success": True,
            "params": body.model_dump(),
            "count": len(result),
            "data": result,
        }
    except HTTPException:
        raise
    except Exception as error:
        print(f"Query error: {error}")
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


# Protected endpoint (requires JWT)
@app.get("/protected")
async def protected(request: Request):
    """
    Protected endpoint that requires authentication.

    This endpoint demonstrates:
    - JWT authentication
    - Accessing JWT claims from the request
    """
    moose = get_moose_utils(request)
    if not moose or not moose.jwt:
        raise HTTPException(
            status_code=401,
            detail="Unauthorized - JWT token required"
        )

    return {
        "message": "You are authenticated",
        "user": moose.jwt.get("sub"),
        "claims": moose.jwt,
    }


# Error handler
@app.exception_handler(Exception)
async def global_exception_handler(request: Request, exc: Exception):
    """Global error handler for unhandled exceptions"""
    print(f"FastAPI error: {exc}")
    return JSONResponse(
        status_code=500,
        content={
            "error": "Internal Server Error",
            "message": str(exc),
        }
    )


# Register as a WebApp with custom mount path
bar_webapp_api = WebApp(
    "barFastAPI",
    app,
    WebAppConfig(
        mount_path="/fastapi",
        metadata=WebAppMetadata(
            description="FastAPI API with middleware demonstrating WebApp integration"
        ),
    )
)
